File: Algorithms/Sort/shell_sort/shell_sort.py
from typing import List


def shell_sort(array: List[int], is_descend: False):
    gap = len(array) // 2
    while gap > 0:
        for i in range(len(array) - gap, len(array)):
            key = array[i]
            j = i - gap
            while j >= 0 and ((array[j] > key) ^ is_descend):
                array[j+gap] = array[j]
                j -= gap
            array[j+gap] = key
        gap //= 2

File: algorithms/sort/shell_sort/shell_sort.cpp
#include "../sort.h"

#include <iostream>

void ShellSort(int* array, uint length, bool isDescend)
{
    int i, j, gap, key;
    for (gap = length / 2; gap > 0; gap /= 2) {
        for (i = gap; i < length; i++) {
            key = array[i];
            for (j = i - gap; j > -1 && ((array[j] > key) ^ isDescend); j -= gap)
                array[j + gap] = array[j];
            array[j + gap] = key;
        }
    }
}